blockprod/src/detail/mod.rs

```diff
@@ -202,41 +171,19 @@ impl BlockProduction {
         Ok(block_header)
     }
 
-    pub fn make_job_key(tip_at_start: &GenBlockIndex) -> JobKey {
-        JobKey {
-            current_tip: tip_at_start.block_id(),
-        }
-    }
-
     pub async fn generate_block(
         &mut self,
         _reward_destination: Destination,
         transactions_source: TransactionsSource,
     ) -> Result<Block, BlockProductionError> {
         let stop_flag = Arc::new(false.into());
 
-        let (cancel_sender, mut cancel_receiver) = oneshot::channel::<()>();
-
         let timestamp = BlockTimestamp::from_duration_since_epoch(self.time_getter().get_time());
 
         let (_, tip_at_start) = self.pull_consensus_data(timestamp).await?;
 
-        {
-            // define the job and insert it into the map of all jobs
-            let job_key = Self::make_job_key(&tip_at_start);
-
-            #[allow(clippy::map_entry)]
-            if !self.all_jobs.contains_key(&job_key) {
-                self.all_jobs.insert(
-                    job_key,
-                    JobHandle {
-                        cancel_signal: cancel_sender,
-                    },
-                );
-            } else {
-                return Err(BlockProductionError::JobAlreadyExists(job_key));
-            }
-        }
+        let (_job_key, mut cancel_receiver) =
+            self.job_manager.new_job(tip_at_start.block_id()).await?;
```

Comment:
I think this function is incomplete. By the end of this function, this key should be removed from the job_manager.

There are multiple ways to do this. 

There's the naive C way of doing this, where we call this function before returning, and hope no one in the future will change the function to return before and cause a memory leak.

There's the smart way to do this, where we create an RAII wrapper, which will work on deleting the job from the queue automatically when this scope finishes.